### ISSUE TITLE
Possible cache-and-network fetchPolicy issue with re-rendering on cache query update if app starts offline

### ISSUE THREAD

**Issue by rgbedin** [apollographql/apollo-client#7623]
**Intended outcome:**
1. Application starts in offline mode
2. Queries loads all available results from cache 
3. Components get updated/rendered using cached state
4. Application tries to fetch queries from network and hangs since client is offline
5. User triggers an action that causes a `useMutation()` to be triggered w/ optimistic response
6. `update()` is called right-away with optimistic object
7. Entry is added to a query cache by using `cache.modify()` for the modified query
8. Component consuming query via `useQuery()` hook gets updated with optimistic response

**Actual outcome:**
If the application is loaded and the client is offline, the `update()` is triggered and the optimistic response is added successfully to the cache (_steps 1-7 are working as expected_); however the UI that uses `useQuery()` for the updated query does not get re-rendered/updated after the cache for the query is updated (_step 8 fails_).

I tried playing around with the `fetchPolicy` on the `defaultOptions` for the client and it seems that if I set it to `cache-first` all the behaviour listed in **Intended outcome** works as expected. Therefore that means that this is most likely to be an issue with `cache-and-network`. 

To add on that, if the application does an initial fetch from the server successfully _and then_ goes offline, it behaves as expected as well - optimistic response added to the query cache successfully triggers an UI update. It seems Apollo wants the first fetch from the server to resolve. 

Maybe this is an issue with `cache-and-network`. However if I change this policy I don't achieve the initial load-from-server that I need.

**Environment:**
- React Native (Non-Expo) v0.63.3
- iOS Simulator and real device iOS 14.2
- Apollo Client v3.3.6
- Using Network Link Conditioner on "100% Loss" profile (iOS Simulator) and Airplane Mode on real device to simulate offline startup

**Apollo Client Configuration:**
```javascript
// Setup links
const httpLink = createHttpLink({
  uri: ENV.GRAPHQL_SERVER_URI
});

const retryLink = new RetryLink({ attempts: { max: Infinity } });

const operationLogLink = new ApolloLink((operation, forward) => {
  console.info(
    `ApolloContext: [${operation.operationName}]:`,
    operation.variables
  );
  operation.setContext({ start: new Date() });

  return forward(operation).map((data) => {
    const duration =
      new Date().valueOf() - operation.getContext().start.valueOf();
    console.info(
      `ApolloContext: [${operation.operationName}] [${duration}ms]:`,
      data
    );

    return data;
  });
});

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.map(({ message, locations, path }) =>
      console.warn(
        `GraphQL error: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  if (networkError) {
    // TODO: display network error UI
    console.warn(`GraphQL network error: ${networkError}`);
  }
});

const authLink = setContext((_, { headers }) => {
  return {
    headers: {
      ...headers,
      authorization: storedSessionRef.current?.accessToken
        ? `Bearer ${storedSessionRef.current?.accessToken}`
        : '',
      orgGuid: storedSessionRef.current?.orgGuid || ''
    }
  };
});

const link = from([
  authLink,
  operationLogLink,
  errorLink,
  retryLink,
  httpLink
]);

// Setup cache and persistor
const cache = apolloCache;

const newPersistor = new CachePersistor({
  cache,
  storage: new AsyncStorageWrapper(AsyncStorage),
  debug: __DEV__,
  trigger: 'write'
});

await newPersistor.restore();
setPersistor(newPersistor);

// Check if we have initial cache
const cacheSize = await newPersistor.getSize();
setHasStoredCache(!!cacheSize && cacheSize > 0);

// Finally, create client
const client = new ApolloClient({
  link,
  queryDeduplication: false,
  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'cache-and-network',
      nextFetchPolicy: 'cache-first',
      errorPolicy: 'all'
    },
    query: {
      errorPolicy: 'all'
    },
    mutate: {
      errorPolicy: 'all'
    }
  },
  cache,
  connectToDevTools: true
});

setClient(client);

console.info('ApolloContext: we are ready:', !!client);
```

**How to reproduce the issue:**
Steps listed in **Intended outcome**.

**Versions**
```
  System:
    OS: macOS 10.15.4
  Binaries:
    Node: 12.18.4 - /usr/local/bin/node
    Yarn: 1.22.5 - /usr/local/bin/yarn
    npm: 6.14.6 - /usr/local/bin/npm
  Browsers:
    Chrome: 88.0.4324.96
    Firefox: 83.0
    Safari: 13.1
  npmPackages:
    @apollo/client: ^3.3.6 => 3.3.6 
    apollo-link-offline: ^1.0.1 => 1.0.1 
    apollo3-cache-persist: ^0.9.1 => 0.9.1 
```

**Comment by rgbedin** [comment]
Mote updates on this: it seems if I also remove the retry link the app works as expected. It might be something that the retry link does in the case there is no initial network available. I tried digging up in the source code of it, but I am not able to detect what could be happening.

**Comment by rgbedin** [comment]
I finally realized what is happening. The `RetryLink` is tricky with "cache-and-network" policy.

If you use the `RetryLink` with a query and "cache-and-network", the logic is:
1. Apollo starts the query
2. Apollo gets the cached data and serves that back
3. Apollo tries to reach the network
4. Apollo replies with updated network data

If the `RetryLink` is into play and a request is started, that means 3 *hangs* until the network is
actually back. That means until the network is actually back, this query data *WILL NOT* get refreshed
into the component. Even if you fire a mutation/change the cache while the query is in-flight, it will not
trigger that the data was updated in the component using the query,

Therefore, as an example, if we start a `GetMessages` and network is offline, the component will render
with the cached data, start the network request and *hang*. If you try to post a message via mutation, even with optimistic
reply that writes to the `GetMessages` cache, it *will not* trigger that the data for the query is updated on
the component. 

Phew.

That means for these type of queries, we need to not use the `RetryLink` and let the query fail with a `NetworkError`.
I added a code to be able to bypass the `RetryLink` by setting 'skipRetryLink' in the query context when creating them.
Then, we set the example `GetMessage` to skip the retry link and fail. 

We should keep in mind that *we are responsible for updating the query manually when network is available*, since the query will not try again by itself.
